File: messages-service.js
/* Supabase-backed messaging service. The conversation list is intentionally
 * lightweight; messages and private attachment URLs load only for the chat
 * that the user opens. */
(function (root) {
  "use strict";

  var CONVERSATION_SELECT = "id, factory_id, individual_id, last_message_at, created_at";
  var LEGACY_CONVERSATION_SELECT =
    CONVERSATION_SELECT + ", factories(name, logo), " +
    "profiles!conversations_individual_id_fkey(full_name, company_image), " +
    "messages(id, sender_id, body, attachment_url, attachment_type, created_at)";
  var MESSAGE_SELECT = "id, sender_id, body, attachment_url, attachment_type, created_at";

  function ready() {
    if (!root.sb || !root.SF_USER) return Promise.reject(new Error("يجب تسجيل الدخول للرسائل"));
    return Promise.resolve();
  }

  function validFactoryId(value) {
    var id = Number(value);
    return Number.isInteger(id) && id > 0 ? id : null;
  }

  function validConversationId(value) {
    var id = Number(value);
    return Number.isInteger(id) && id > 0 ? id : null;
  }

  function messageTime(message) {
    var time = new Date(message.created_at).getTime();
    return isNaN(time) ? 0 : time;
  }

  function mapMessage(message, currentId) {
    return {
      id: message.id,
      from: message.sender_id === currentId ? "mine" : "theirs",
      type: message.attachment_type || "text",
      text: message.body || "",
      src: message.signed_url || "",
      at: messageTime(message)
    };
  }

  function signMessages(messages) {
    var jobs = [];

    (messages || []).forEach(function (message) {
      if (!message.attachment_url) return;
      jobs.push(
        root.sb.storage.from("chat-media")
          .createSignedUrl(message.attachment_url, 3600)
          .then(function (res) {
            if (!res.error && res.data) message.signed_url = res.data.signedUrl;
          })
          .catch(function () {
            /* تبقى الرسالة النصية ظاهرة حتى لو تعذّر توقيع المرفق مؤقتًا. */
          })
      );
    });

    return Promise.all(jobs).then(function () { return messages || []; });
  }

  function addPeerMetadata(rows) {
    return root.sb.rpc("get_conversation_peers").then(function (res) {
      if (res.error) return rows || [];

      var peers = {};
      (res.data || []).forEach(function (peer) {
        peers[String(peer.conversation_id)] = peer;
      });
      (rows || []).forEach(function (row) {
        row.peer = peers[String(row.id)] || null;
      });
      return rows || [];
    }).catch(function () {
      return rows || [];
    });
  }

  function summaryToThread(row) {
    var lastAt = row.last_message_created_at || row.last_message_at;
    var updated = new Date(lastAt || 0).getTime();
    return {
      id: String(row.conversation_id),
      conversation_id: row.conversation_id,
      factory_id: row.factory_id,
      name: row.peer_name || "",
      role: row.peer_role || "",
      avatar: row.peer_avatar || "",
      messages: [],
      messagesLoaded: false,
      lastMessage: row.last_message_created_at ? {
        from: row.last_message_sender_id === root.SF_USER.id ? "mine" : "theirs",
        type: row.last_message_type || "text",
        text: row.last_message_body || "",
        at: new Date(row.last_message_created_at).getTime()
      } : null,
      unread: Math.max(0, Number(row.unread_count || 0)),
      updated: isNaN(updated) ? 0 : updated
    };
  }

  function legacyToThread(row, currentId) {
    var factory = row.factories || {};
    var individual = row.profiles || {};
    var isFactory = root.SF_USER && row.individual_id !== root.SF_USER.id;
    var peer = row.peer || {};
    var messages = (row.messages || []).slice().sort(function (a, b) {
      return messageTime(a) - messageTime(b) || Number(a.id || 0) - Number(b.id || 0);
    }).map(function (message) {
      return mapMessage(message, currentId);
    });
    return {
      id: String(row.id),
      conversation_id: row.id,
      factory_id: row.factory_id,
      name: peer.peer_name || (isFactory ? (individual.full_name || "") : (factory.name || "")),
      role: peer.peer_role || (isFactory ? "individual" : "factory"),
      avatar: peer.peer_avatar || (isFactory ? (individual.company_image || "") : (factory.logo || "")),
      messages: messages,
      messagesLoaded: true,
      lastMessage: messages.length ? messages[messages.length - 1] : null,
      unread: 0,
      updated: new Date(row.last_message_at || row.created_at).getTime()
    };
  }

  function shellToThread(row) {
    var peer = row.peer || {};
    return {
      id: String(row.id),
      conversation_id: row.id,
      factory_id: row.factory_id,
      name: peer.peer_name || "",
      role: peer.peer_role || "",
      avatar: peer.peer_avatar || "",
      messages: [],
      messagesLoaded: false,
      lastMessage: null,
      unread: 0,
      updated: new Date(row.last_message_at || row.created_at).getTime()
    };
  }

  function loadLegacy() {
    return root.sb.from("conversations")
      .select(LEGACY_CONVERSATION_SELECT)
      .order("last_message_at", { ascending: false })
      .then(function (res) {
        if (res.error) throw res.error;
        return addPeerMetadata(res.data || []);
      }).then(function (rows) {
        var rawMessages = [];
        rows.forEach(function (row) {
          (row.messages || []).forEach(function (message) { rawMessages.push(message); });
        });
        return signMessages(rawMessages).then(function () { return rows; });
      }).then(function (rows) {
        var out = {};
        rows.forEach(function (row) {
          out[String(row.id)] = legacyToThread(row, root.SF_USER.id);
        });
        return out;
      });
  }

  function threadFromRow(row) {
    return addPeerMetadata([row]).then(function (rows) {
      return shellToThread(rows[0]);
    });
  }

  function findConversation(factoryId) {
    return root.sb.from("conversations")
      .select(CONVERSATION_SELECT)
      .eq("factory_id", factoryId)
      .eq("individual_id", root.SF_USER.id)
      .maybeSingle();
  }

  function insertConversation(factoryId) {
    return root.sb.from("conversations").insert({
      factory_id: factoryId,
      individual_id: root.SF_USER.id
    }).select(CONVERSATION_SELECT).single();
  }

  root.SFMessages = {
    load: function () {
      return ready().then(function () {
        return root.sb.rpc("get_conversation_summaries");
      }).then(function (res) {
        if (res.error) return loadLegacy();

        var out = {};
        (res.data || []).forEach(function (row) {
          out[String(row.conversation_id)] = summaryToThread(row);
        });
        return out;
      });
    },

    loadMessages: function (conversationId) {
      return ready().then(function () {
        var id = validConversationId(conversationId);
        if (!id) throw new Error("معرّف المحادثة غير صالح");
        return root.sb.from("messages")
          .select(MESSAGE_SELECT)
          .eq("conversation_id", id)
          .order("created_at", { ascending: true })
          .order("id", { ascending: true });
      }).then(function (res) {
        if (res.error) throw res.error;
        return signMessages(res.data || []);
      }).then(function (rows) {
        return rows.map(function (message) {
          return mapMessage(message, root.SF_USER.id);
        });
      });
    },

    markRead: function (conversationId) {
      return ready().then(function () {
        var id = validConversationId(conversationId);
        if (!id) throw new Error("معرّف المحادثة غير صالح");
        return root.sb.rpc("mark_conversation_read", { p_conversation_id: id });
      }).then(function (res) {
        if (res.error) {
          /* توافق مؤقت قبل تطبيق ترحيل العدّاد. */
          return root.sb.from("messages")
            .update({ read_at: new Date().toISOString() })
            .eq("conversation_id", Number(conversationId))
            .neq("sender_id", root.SF_USER.id)
            .is("read_at", null)
            .then(function (fallback) {
              if (fallback.error) throw fallback.error;
              return 0;
            });
        }
        return Number(res.data || 0);
      });
    },

    listFactories: function () {
      return ready().then(function () {
        return root.sb.from("factories")
          .select("id, name, logo")
          .eq("status", "approved")
          .order("name", { ascending: true });
      }).then(function (res) {
        if (res.error) throw res.error;
        return res.data || [];
      });
    },

    ensureFactoryConversation: function (factoryId) {
      return ready().then(function () {
        var id = validFactoryId(factoryId);
        if (!id) throw new Error("معرّف المصنع غير صالح");
        if (root.SF_PROFILE && root.SF_PROFILE.account_type === "factory") {
          throw new Error("حساب المصنع يستقبل محادثات العملاء ولا يبدأ محادثة مع مصنع آخر");
        }
        return findConversation(id).then(function (res) {
          return { id: id, response: res };
        });
      }).then(function (state) {
        var res = state.response;
        if (res.error) throw res.error;
        if (res.data) return threadFromRow(res.data);

        return insertConversation(state.id).then(function (created) {
          if (!created.error) return threadFromRow(created.data);
          if (created.error.code === "23505") {
            return findConversation(state.id).then(function (retry) {
              if (retry.error) throw retry.error;
              if (!retry.data) throw created.error;
              return threadFromRow(retry.data);
            });
          }
          throw created.error;
        });
      });
    },

    sendAttachment: function (conversationId, file, type) {
      return ready().then(function () {
        var extension = (file.name || "bin").split(".").pop().toLowerCase().replace(/[^a-z0-9]/g, "") || "bin";
        var path = root.SF_USER.id + "/" + (root.crypto && root.crypto.randomUUID ? root.crypto.randomUUID() : String(Date.now())) + "." + extension;
        return root.sb.storage.from("chat-media").upload(path, file, { upsert: false, contentType: file.type || undefined });
      }).then(function (uploadRes) {
        if (uploadRes.error) throw uploadRes.error;
        return root.sb.from("messages").insert({
          conversation_id: Number(conversationId),
          sender_id: root.SF_USER.id,
          body: "",
          attachment_url: uploadRes.data.path,
          attachment_type: type
        }).select(MESSAGE_SELECT).single();
      }).then(function (messageRes) {
        if (messageRes.error) throw messageRes.error;
        return signMessages([messageRes.data]).then(function (rows) { return rows[0]; });
      });
    },

    sendText: function (conversationId, text) {
      return ready().then(function () {
        return root.sb.from("messages").insert({
          conversation_id: Number(conversationId),
          sender_id: root.SF_USER.id,
          body: String(text || "").slice(0, 10000),
          attachment_url: "",
          attachment_type: ""
        }).select(MESSAGE_SELECT).single();
      }).then(function (res) {
        if (res.error) throw res.error;
        return res.data;
      });
    },

    subscribe: function (onChange) {
      if (!root.sb || !root.SF_USER) return null;
      return root.sb.channel("sf-messages-" + root.SF_USER.id)
        .on("postgres_changes", { event: "*", schema: "public", table: "messages" }, onChange)
        .subscribe();
    }
  };
}(window));

File: app_flutter/lib/services/messages_service.dart
// ============================================================
//  خدمة الرسائل — مقابل messages-service.js
//
//  قائمة المحادثات تُبنى من دوال الخادم (RPC) لا بربط الجداول
//  في العميل: سياسة RLS ترفض قراءة صفوف profiles لطرف آخر،
//  ولذلك وُجدت get_conversation_peers.
//
//  المرفقات في دلو خاص، فتُوقَّع روابطها (signed URL) لساعة.
// ============================================================

import 'dart:async';

import 'package:supabase_flutter/supabase_flutter.dart';

import '../core/supabase_config.dart';
import '../core/uuid.dart';
import 'auth_service.dart';
import 'chat_upload_service.dart';
import 'factory_service.dart';

const String _conversationSelect =
    'id, factory_id, individual_id, last_message_at, created_at';
const String _messageSelect =
    'id, sender_id, body, attachment_url, attachment_type, product_id, client_key, created_at';

/// بيانات البطاقة تأتي من RPC المحادثة لتبقى متاحة للطرفين.
class SFProductAttachment {
  const SFProductAttachment({
    required this.id,
    this.name = '',
    this.image = '',
    this.price = 0,
  });
  final int id;
  final String name;
  final String image;
  final double price;

  factory SFProductAttachment.fromRow(Map<String, dynamic> row) =>
      SFProductAttachment(
        id: (row['product_id'] as num).toInt(),
        name: row['name'] as String? ?? '',
        image: row['image'] as String? ?? '',
        price: SFProduct({...row, 'id': row['product_id']}).minPrice,
      );
}

/// رسالة واحدة بعد التطبيع — نفس شكل mapMessage في الويب.
class SFMessage {
  SFMessage({
    required this.id,
    required this.mine,
    required this.type,
    required this.text,
    required this.src,
    required this.at,
    this.pending = false,
    this.failed = false,
    this.clientKey = '',
    this.product,
    this.productId,
  });

  final String id;

  /// من المستخدم الحالي؟ (mine مقابل theirs في الويب)
  final bool mine;

  /// text أو image أو video.
  final String type;
  final String text;

  /// رابط المرفق الموقّع.
  final String src;
  final DateTime at;

  /// رسالة متفائلة لم يؤكّدها الخادم بعد.
  final bool pending;
  final bool failed;
  final String clientKey;
  final SFProductAttachment? product;
  final int? productId;

  bool get isText => type == 'text' || type.isEmpty;

  SFMessage copyWith({bool? pending, bool? failed}) => SFMessage(
    id: id,
    mine: mine,
    type: type,
    text: text,
    src: src,
    at: at,
    pending: pending ?? this.pending,
    failed: failed ?? this.failed,
    clientKey: clientKey,
    product: product,
    productId: productId,
  );

  static SFMessage fromRow(
    Map<String, dynamic> m,
    String currentId, {
    String signedUrl = '',
    SFProductAttachment? product,
  }) {
    return SFMessage(
      id: '${m['id']}',
      mine: m['sender_id'] == currentId,
      type: (m['attachment_type'] as String?)?.isNotEmpty == true
          ? m['attachment_type'] as String
          : 'text',
      text: (m['body'] as String?) ?? '',
      src: signedUrl,
      product: product,
      productId: (m['product_id'] as num?)?.toInt(),
      clientKey: (m['client_key'] as String?) ?? '',
      at:
          DateTime.tryParse('${m['created_at']}')?.toLocal() ??
          DateTime.fromMillisecondsSinceEpoch(0),
    );
  }
}

/// محادثة في القائمة.
class SFThread {
  SFThread({
    required this.id,
    required this.conversationId,
    required this.factoryId,
    required this.name,
    required this.role,
    required this.avatar,
    required this.updated,
    this.unread = 0,
    this.lastMessage,
    this.messages = const [],
    this.messagesLoaded = false,
  });

  final String id;
  final int conversationId;
  final int? factoryId;
  final String name;
  final String role;
  final String avatar;
  final DateTime updated;
  final int unread;
  final SFMessage? lastMessage;
  final List<SFMessage> messages;
  final bool messagesLoaded;

  SFThread copyWith({
    List<SFMessage>? messages,
    bool? messagesLoaded,
    int? unread,
    SFMessage? lastMessage,
    DateTime? updated,
  }) => SFThread(
    id: id,
    conversationId: conversationId,
    factoryId: factoryId,
    name: name,
    role: role,
    avatar: avatar,
    updated: updated ?? this.updated,
    unread: unread ?? this.unread,
    lastMessage: lastMessage ?? this.lastMessage,
    messages: messages ?? this.messages,
    messagesLoaded: messagesLoaded ?? this.messagesLoaded,
  );
}

class SFMessages {
  SFMessages._();

  static int _subscriptionSerial = 0;

  static void _ready() {
    if (AuthService.instance.user == null) {
      throw Exception('يجب تسجيل الدخول للرسائل');
    }
  }

  static String get _uid => AuthService.instance.user!.id;

  /// قائمة المحادثات — عبر get_conversation_summaries.
  static Future<List<SFThread>> load() async {
    _ready();
    final rows = await sb.rpc('get_conversation_summaries') as List<dynamic>?;
    final list = (rows ?? []).map((r) {
      final m = r as Map<String, dynamic>;
      final lastAt = m['last_message_created_at'] ?? m['last_message_at'];
      final updated =
          DateTime.tryParse('$lastAt')?.toLocal() ??
          DateTime.fromMillisecondsSinceEpoch(0);
      SFMessage? last;
      if (m['last_message_created_at'] != null) {
        last = SFMessage(
          id: 'last-${m['conversation_id']}',
          mine: m['last_message_sender_id'] == _uid,
          type: (m['last_message_type'] as String?)?.isNotEmpty == true
              ? m['last_message_type'] as String
              : 'text',
          text: (m['last_message_body'] as String?) ?? '',
          src: '',
          at: updated,
        );
      }
      return SFThread(
        id: '${m['conversation_id']}',
        conversationId: (m['conversation_id'] as num).toInt(),
        factoryId: (m['factory_id'] as num?)?.toInt(),
        name: (m['peer_name'] as String?) ?? '',
        role: (m['peer_role'] as String?) ?? '',
        avatar: (m['peer_avatar'] as String?) ?? '',
        updated: updated,
        unread: ((m['unread_count'] as num?)?.toInt() ?? 0).clamp(0, 1 << 30),
        lastMessage: last,
      );
    }).toList();
    list.sort((a, b) => b.updated.compareTo(a.updated));
    return list;
  }

  /// رسائل محادثة واحدة، مع توقيع روابط المرفقات.
  static Future<List<SFMessage>> loadMessages(int conversationId) async {
    _ready();
    if (conversationId <= 0) throw Exception('معرّف المحادثة غير صالح');
    final rows = await sb
        .from('messages')
        .select(_messageSelect)
        .eq('conversation_id', conversationId)
        .order('created_at', ascending: true)
        .order('id', ascending: true);

    final products = <int, SFProductAttachment>{};
    if (rows.any((row) => row['product_id'] != null)) {
      try {
        final cards = await sb.rpc(
          'get_message_products',
          params: {'p_conversation_id': conversationId},
        );
        for (final row in cards as List) {
          final product = SFProductAttachment.fromRow(
            Map<String, dynamic>.from(row as Map),
          );
          products[product.id] = product;
        }
      } catch (_) {
        // فشل البطاقة لا يُخفي نصوص المحادثة.
      }
    }
    final out = <SFMessage>[];
    for (final r in rows) {
      final m = Map<String, dynamic>.from(r as Map);
      final path = (m['attachment_url'] as String?) ?? '';
      String signed = '';
      if (path.isNotEmpty) {
        try {
          signed = await sb.storage
              .from(SFBuckets.chatMedia)
              .createSignedUrl(path, 3600);
        } catch (_) {
          // تبقى الرسالة ظاهرة حتى لو تعذّر توقيع المرفق.
        }
      }
      out.add(
        SFMessage.fromRow(
          m,
          _uid,
          signedUrl: signed,
          product: products[(m['product_id'] as num?)?.toInt()],
        ),
      );
    }
    return out;
  }

  /// تعليم المحادثة مقروءة — الدالة ترفض غير الطرفين بـ 42501.
  static Future<int> markRead(int conversationId) async {
    _ready();
    if (conversationId <= 0) throw Exception('معرّف المحادثة غير صالح');
    try {
      final res = await sb.rpc(
        'mark_conversation_read',
        params: {'p_conversation_id': conversationId},
      );
      return (res as num?)?.toInt() ?? 0;
    } catch (_) {
      return 0;
    }
  }

  /// إجمالي غير المقروء — لشارة الرسائل في الشريط السفلي.
  static Future<int> unreadTotal() async {
    if (AuthService.instance.user == null) return 0;
    try {
      final res = await sb.rpc('get_unread_message_total');
      return (res as num?)?.toInt() ?? 0;
    } catch (_) {
      return 0;
    }
  }

  static Future<List<Map<String, dynamic>>> listFactories() async {
    _ready();
    final rows = await sb
        .from('factories')
        .select('id, name, logo')
        .eq('status', 'approved')
        .order('name', ascending: true);
    return rows.map((e) => Map<String, dynamic>.from(e as Map)).toList();
  }

  /// يفتح محادثة مع مصنع أو ينشئها.
  static Future<SFThread> ensureFactoryConversation(int factoryId) async {
    _ready();
    if (factoryId <= 0) throw Exception('معرّف المصنع غير صالح');
    if (AuthService.instance.profile?.accountType == 'factory') {
      throw Exception(
        'حساب المصنع يستقبل محادثات العملاء ولا يبدأ محادثة مع مصنع آخر',
      );
    }

    Map<String, dynamic>? row = await sb
        .from('conversations')
        .select(_conversationSelect)
        .eq('factory_id', factoryId)
        .eq('individual_id', _uid)
        .maybeSingle();

    if (row == null) {
      try {
        row = await sb
            .from('conversations')
            .insert({'factory_id': factoryId, 'individual_id': _uid})
            .select(_conversationSelect)
            .single();
      } on PostgrestException catch (e) {
        // 23505: أنشأها طلب متزامن — نقرأها بدل الفشل.
        if (e.code != '23505') rethrow;
        row = await sb
            .from('conversations')
            .select(_conversationSelect)
            .eq('factory_id', factoryId)
            .eq('individual_id', _uid)
            .maybeSingle();
        if (row == null) rethrow;
      }
    }

    final peers = await _peers();
    final peer = peers['${row['id']}'];
    return SFThread(
      id: '${row['id']}',
      conversationId: (row['id'] as num).toInt(),
      factoryId: (row['factory_id'] as num?)?.toInt(),
      name: (peer?['peer_name'] as String?) ?? '',
      role: (peer?['peer_role'] as String?) ?? '',
      avatar: (peer?['peer_avatar'] as String?) ?? '',
      updated:
          DateTime.tryParse('${row['last_message_at'] ?? row['created_at']}')
              ?.toLocal() ??
          DateTime.now(),
    );
  }

  static Future<Map<String, Map<String, dynamic>>> _peers() async {
    try {
      final res = await sb.rpc('get_conversation_peers') as List<dynamic>?;
      final out = <String, Map<String, dynamic>>{};
      for (final r in res ?? []) {
        final m = Map<String, dynamic>.from(r as Map);
        out['${m['conversation_id']}'] = m;
      }
      return out;
    } catch (_) {
      return {};
    }
  }

  static Future<SFMessage> sendText(
    int conversationId,
    String text, {
    String? clientKey,
    SFProductAttachment? product,
  }) async {
    _ready();
    if (text.trim().isEmpty && product == null) {
      throw Exception('اكتب رسالة أولاً');
    }
    final key = clientKey ?? newUuidV4();
    final row = await sb
        .from('messages')
        .insert({
          'conversation_id': conversationId,
          'sender_id': _uid,
          'body': text.length > 10000 ? text.substring(0, 10000) : text,
          'attachment_url': '',
          'attachment_type': product == null ? '' : 'product',
          if (product != null) 'product_id': product.id,
          'client_key': key,
        })
        .select(_messageSelect)
        .single();
    return SFMessage.fromRow(
      Map<String, dynamic>.from(row),
      _uid,
      product: product,
    );
  }

  static Future<SFMessage> sendMedia(
    int conversationId,
    SFChatUpload upload, {
    required String clientKey,
  }) async {
    _ready();
    if (!upload.path.startsWith('$_uid/') ||
        !['image', 'video'].contains(upload.type)) {
      throw StateError('مرفق غير صالح');
    }
    final row = await sb
        .from('messages')
        .insert({
          'conversation_id': conversationId,
          'sender_id': _uid,
          'body': '',
          'attachment_url': upload.path,
          'attachment_type': upload.type,
          'client_key': clientKey,
        })
        .select(_messageSelect)
        .single();
    return SFMessage.fromRow(
      Map<String, dynamic>.from(row),
      _uid,
      signedUrl: upload.url,
    );
  }

  /// يشترك في تغيّرات جدول الرسائل — البديل عن realtime في الويب.
  static RealtimeChannel? subscribe(
    void Function(PostgresChangePayload payload) onChange,
  ) {
    if (AuthService.instance.user == null) return null;
    final serial = _subscriptionSerial++;
    return sb
        // يجب أن يكون موضوع Phoenix فريداً لكل اشتراك نشط؛ فالهيكل وقائمة
        // المحادثات وصفحة المحادثة قد تكون مشتركة في الوقت نفسه.
        .channel('sf-messages-$_uid-$serial')
        .onPostgresChanges(
          event: PostgresChangeEvent.all,
          schema: 'public',
          table: 'messages',
          callback: onChange,
        )
        .subscribe();
  }
}
